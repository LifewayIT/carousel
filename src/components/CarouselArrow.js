import React from 'react';
import styled from 'styled-components';
import { color, shadow, device, space } from '../utils/styleguide';

import ChevronLeft from '../../svg/chevron-left.svg';
import ChevronRight from '../../svg/sprite/chevron-right.svg';

const Arrow = styled.button`
  position: absolute;
  display: block;

  padding: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 50%;
  border: 0 none;
  
  color: ${color.gray800};
  background-color: ${color.gray100};
  box-shadow:
    ${shadow.button},
    inset 0 0 4px 0 rgb(0 0 0 / 0.05);

  @media (prefers-reduced-motion: no-preference) {
    transition: all .1s ease;
  }
  
  &:not(:disabled):hover {
    cursor: pointer;
    box-shadow:
      ${shadow.buttonHover},
      inset 0 0 4px 0 rgb(0 0 0 / 0.05);
    transform: scale(1.05) translateY(-2px);
  }

  @media (hover: none), (hover: on-demand) {
    &:not(:disabled):hover {
      box-shadow:
        ${shadow.button},
        inset 0 0 4px 0 rgb(0 0 0 / 0.05); 
      transform: unset;
    }
  }

  &:not(:disabled):active {
    background: ${color.gray200};
  }

  &:disabled {
    opacity: .6;
  }

  ${props => props.hide && `
    visibility: hidden;
    transform: scale(0);

    @media (prefers-reduced-motion: no-preference) {
      transition: all .15s ease-in;
    }
  `}

  width: 48px;
  height: 48px;

  & svg, & img {
    width: 16px;
    height: 16px;
  }

  top: calc(50% - 24px);
  ${props => props.left && `left: ${space._8};`}
  ${props => props.right && `right: ${space._8};`}
  z-index: 1;


  @media ${device._768} {
    width: 64px;
    height: 64px;

    & svg, & img {
      width: 24px;
      height: 24px;
    }

    top: calc(50% - 32px);
    ${props => props.left && `left: ${space._16};`}
    ${props => props.right && `right: ${space._16};`}
  }
`;

const CarouselArrow = ({ left, right, onClick, ...props }) => {
  if (!left && !right) {
    throw new Error('either "left" or "right" prop is required');
  } else if (left && right) {
    throw new Error('cannot have both "left" and "right" props');
  }

  return (
    <Arrow left={left} right={right} onClick={onClick} {...props}>
      {left && <img src={ChevronLeft} alt="page left" />}
      {right && <img src={ChevronRight} alt="page right" />}
    </Arrow>
  );
};

export default CarouselArrow;
