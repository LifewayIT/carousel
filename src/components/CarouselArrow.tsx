import React, { ButtonHTMLAttributes, MouseEventHandler, ReactElement } from 'react';
import styled from 'styled-components';
import { color, device, shadow, space } from '../utils/styleguide';
import { cn } from '../utils/classnames';

import ChevronLeft from '../../svg/chevron-left.svg';
import ChevronRight from '../../svg/chevron-right.svg';


type Props = {
  /** shows the left arrow icon. One of 'left' or 'right' must be specified */
  left?: boolean;
  /** shows the right arrow icon. One of 'left' or 'right' must be specified */
  right?: boolean;
  /** hide the arrow */
  hide?: boolean;
  /** handler for when the arrow button is clicked */
  onClick: MouseEventHandler<HTMLElement>;
} & ButtonHTMLAttributes<HTMLButtonElement>;

export const CarouselArrow = ({ left, right, onClick, className, ...props }: Props): ReactElement => {
  if (!left && !right) {
    throw new Error('either "left" or "right" prop is required');
  } else if (left && right) {
    throw new Error('cannot have both "left" and "right" props');
  }

  const cls = cn('lwc-arrow', { 'lwc-left': left, 'lwc-right': right }, className);

  return (
    <Arrow {...props} className={cls} onClick={onClick}>
      {left && <img src={ChevronLeft} alt="page left" />}
      {right && <img src={ChevronRight} alt="page right" />}
    </Arrow>
  );
};


type ArrowProps = {
  left?: boolean;
  right?: boolean;
  hide?: boolean;
};

const Arrow = styled.button<ArrowProps>`
  position: absolute;
  display: block;

  padding: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 50%;
  border: 0 none;
  
  color: ${color.gray800};
  background-color: ${color.gray100};
  box-shadow:
    ${shadow.button},
    inset 0 0 4px 0 rgb(0 0 0 / 0.05);

  @media (prefers-reduced-motion: no-preference) {
    transition: all .1s ease;
  }
  
  &:not(:disabled):hover {
    cursor: pointer;
    box-shadow:
      ${shadow.buttonHover},
      inset 0 0 4px 0 rgb(0 0 0 / 0.05);
    transform: scale(1.05) translateY(-2px);
  }

  @media (hover: none), (hover: on-demand) {
    &:not(:disabled):hover {
      box-shadow:
        ${shadow.button},
        inset 0 0 4px 0 rgb(0 0 0 / 0.05); 
      transform: unset;
    }
  }

  &:not(:disabled):active {
    background: ${color.gray200};
  }

  &:disabled {
    opacity: .6;
  }

  ${props => props.hide && `
    visibility: hidden;
    transform: scale(0);

    @media (prefers-reduced-motion: no-preference) {
      transition: all .15s ease-in;
    }
  `}

  width: 48px;
  height: 48px;

  & svg, & img {
    width: 16px;
    height: 16px;
  }

  top: calc(50% - 24px);
  &.lwc-left {
    left: ${space._8};
  }
  &.lwc-right {
    right: ${space._8};
  }
  z-index: 1;


  @media ${device._768} {
    width: 64px;
    height: 64px;

    & svg, & img {
      width: 24px;
      height: 24px;
    }

    top: calc(50% - 32px);

    &.lwc-left {
      left: ${space._16};
    }
    &.lwc-right {
      right: ${space._16};
    }
  }
`;
