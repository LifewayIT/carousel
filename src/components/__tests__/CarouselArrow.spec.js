import React from 'react';
import { render } from '@testing-library/react';
import { CarouselArrow } from '../CarouselArrow';

test('left arrow matches snapshot', () => {
  const { container } = render(
    <CarouselArrow className="test-class" left>
      <div id="children" />
    </CarouselArrow>
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('right arrow matches snapshot', () => {
  const { container } = render(
    <CarouselArrow className="test-class" right>
      <div id="children" />
    </CarouselArrow>
  );

  expect(container.firstChild).toMatchSnapshot();
});
