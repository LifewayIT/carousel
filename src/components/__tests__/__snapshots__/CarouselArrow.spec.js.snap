// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`left arrow matches snapshot 1`] = `
<button
  class="lwc-arrow lwc-left test-class"
>
  <img
    alt="page left"
    src="test-file-stub"
  />
</button>
`;

exports[`right arrow matches snapshot 1`] = `
<button
  class="lwc-arrow lwc-right test-class"
>
  <img
    alt="page right"
    src="test-file-stub"
  />
</button>
`;
